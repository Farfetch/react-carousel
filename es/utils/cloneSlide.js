function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import { cloneElement } from 'react';
import Styles from "../components/Slider.css";
import cx from 'classnames';
import memoize from 'fast-memoize';

var getElementStyle = function getElementStyle(style, styleChild) {
  return _objectSpread(_objectSpread({}, style), styleChild);
};

var memoizedGetElementStyle = memoize(getElementStyle);

var cloneSlide = function cloneSlide(child, index, isCloned, className, style) {
  var newStyle = memoizedGetElementStyle(style, child.props.style);
  return /*#__PURE__*/cloneElement(child, {
    className: cx(className, Styles.slide, child.props.className),
    style: newStyle,
    key: isCloned ? "c.".concat(index) : child.key,
    ref: function ref(node) {
      // Keep ref when use cloneElement - Dan Abramov's tweet:
      // https://twitter.com/dan_abramov/status/752944645018189825?lang=en
      if (child.ref && !isCloned) {
        child.ref(node);
      }
    }
  });
};

export default cloneSlide;