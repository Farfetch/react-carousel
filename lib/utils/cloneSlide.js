"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = require("react");

var _Slider = _interopRequireDefault(require("../components/Slider.css"));

var _classnames = _interopRequireDefault(require("classnames"));

var _fastMemoize = _interopRequireDefault(require("fast-memoize"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var getElementStyle = function getElementStyle(style, styleChild) {
  return _objectSpread(_objectSpread({}, style), styleChild);
};

var memoizedGetElementStyle = (0, _fastMemoize["default"])(getElementStyle);

var cloneSlide = function cloneSlide(child, index, isCloned, className, style) {
  var newStyle = memoizedGetElementStyle(style, child.props.style);
  return /*#__PURE__*/(0, _react.cloneElement)(child, {
    className: (0, _classnames["default"])(className, _Slider["default"].slide, child.props.className),
    style: newStyle,
    key: isCloned ? "c.".concat(index) : child.key,
    ref: function ref(node) {
      // Keep ref when use cloneElement - Dan Abramov's tweet:
      // https://twitter.com/dan_abramov/status/752944645018189825?lang=en
      if (child.ref && !isCloned) {
        child.ref(node);
      }
    }
  });
};

var _default = cloneSlide;
exports["default"] = _default;
module.exports = exports.default;