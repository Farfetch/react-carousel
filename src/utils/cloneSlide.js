import { cloneElement } from 'react';
import cx from 'classnames';
import memoize from 'fast-memoize';

const getElementStyle = function (style, styleChild) {
    return { ...style, ...styleChild };
};

const memoizedGetElementStyle = memoize(getElementStyle);

const cloneSlide = (child, index, isCloned, className, style) => {
    const newStyle = memoizedGetElementStyle(style, child.props.style);

    return cloneElement(child, {
        className: cx(className, 'slide', child.props.className),
        style: newStyle,
        key: isCloned ? `c.${index}` : child.key,
        ref: (node) => {
            // Keep ref when use cloneElement - Dan Abramov's tweet:
            // https://twitter.com/dan_abramov/status/752944645018189825?lang=en
            if (child.ref && !isCloned) {
                child.ref(node);
            }
        },
    });
};

export default cloneSlide;
