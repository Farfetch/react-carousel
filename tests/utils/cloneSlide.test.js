import { createSerializer } from 'enzyme-to-json';
import { shallow } from 'enzyme';
import React from 'react';

import cloneSlide from '../../src/utils/cloneSlide';

describe('cloneSlide()', () => {
    beforeAll(() => {
        expect.addSnapshotSerializer(createSerializer());
    });

    it('should return slide element with minimum configuration', () => {
        const child = <div>First slide</div>;

        const slide = cloneSlide(child);

        const wrapper = shallow(slide);

        expect(wrapper).toMatchSnapshot();
    });

    it('should return slide element with additional configuration', () => {
        const child = <div key="0">First slide</div>;

        const dummyStyle = {
            position: 'relative',
        };

        const slide = cloneSlide(child, 1, false, 'dummyClassName', dummyStyle);

        const wrapper = shallow(slide);

        expect(wrapper).toMatchSnapshot();
    });

    it('should return slide element with new key when it is cloned', () => {
        const child = <div>First slide</div>;

        const slide = cloneSlide(child, 1, true);

        const wrapper = shallow(slide);

        expect(wrapper).toMatchSnapshot();
    });
});
