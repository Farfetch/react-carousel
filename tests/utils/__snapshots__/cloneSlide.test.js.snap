// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cloneSlide() should return slide element with additional configuration 1`] = `
<div
  key="0"
  style={
    Object {
      "position": "relative",
    }
  }
>
  First slide
</div>
`;

exports[`cloneSlide() should return slide element with minimum configuration 1`] = `
<div
  key="null"
  style={Object {}}
>
  First slide
</div>
`;

exports[`cloneSlide() should return slide element with new key when it is cloned 1`] = `
<div
  key="c.1"
  style={Object {}}
>
  First slide
</div>
`;
